sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteNonAggregateFirstSuite.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.aggregate.First
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical._
+
+class RewriteNonAggregateFirstSuite extends PlanTest {
+  val testRelation: LocalRelation = LocalRelation('a.string, 'b.string)
+
+  private def checkRewrite(rewrite: LogicalPlan): Unit = rewrite match {
+    case GlobalLimit(_, LocalLimit(_, Project(_, _))) =>
+    case _ => fail(s"Plan is not rewritten:\n$rewrite")
+  }
+
+  test("no group by and single first aggregate") {
+    val input = testRelation.select(first('a)).analyze
+    val rewrite = RewriteNonAggregateFirst(input)
+    checkRewrite(rewrite)
+  }
+
+  test("single first aggregate with group by") {
+    val input = testRelation
+      .groupBy('a)(first('b))
+      .analyze
+    val rewrite = RewriteNonAggregateFirst(input)
+    comparePlans(input, rewrite)
+  }
+
+  test("multiple aggregates with group by") {
+    val input = testRelation
+      .groupBy('a)(
+        first('a').as('agg1),
+        max('b).as('agg2))
+      .analyze
+    val rewrite = RewriteNonAggregateFirst(input)
+    comparePlans(input, rewrite)
+  }
+
+  test("multiple aggregates without group by") {
+    val input = testRelation
+      .select(first('a).as('agg1), max('b).as('agg2))
+      .analyze
+    val rewrite = RewriteNonAggregateFirst(input)
+    comparePlans(input, rewrite)
+  }
+
+  test("no group by and single first aggregate with ignore nulls") {
+    val input = testRelation
+      .select(First('a, ignoreNulls = true).toAggregateExpression())
```

Comment:
add another test with multiple First aggregates (to show the optimization applies), plus one with multiple First aggregates plus one non-First aggregate (to show the optimization does not apply)?

Reply:
+1

Reply:
Done